don't show undescribed enum in combo

dterrahe opened the pull request:
Fixes #13971 but be CAREFUL (see below).

This also tries to remove _one_ reason why a "_LAST" enum sometimes is needed, namely to check if a legacy value that by default is made unavailable needs to be added back when loading an image that uses it. This new approach checks in `dt_bauhaus_combobox_set_from_value` if a non-existing value is being set, in which case it tries to add it back from introspection.

Then also using this to "rebuild" some combos that are tailored in their `reload_defaults`.

This stuff is very tricky to test without "legacy images" and images that tick all the "special case" boxes. So I have basically done _NO_ testing on that; please fill that void (for all the modules touched in this PR).

I would like some actual code review on this as well, not just for "correctness" but also if this actually helps make the "API" more logical or what could be better alternatives. It sort of makes sense to me that some not always available options would be listed at the end (which also makes removing and adding them easy) but in for example the lens correction module this is intentionally not the case, requiring a full rebuild of the list of options.

Maybe an additional `dt_bauhaus_combobox_remove_value` (rather than `dt_bauhaus_combobox_remove_at`) and having `dt_bauhaus_combobox_set_from_value` try to insert at the same location could be worth it, but at the moment there are only very limited use cases so no real justification to add too much complexity (that wouldn't receive sufficient testing).

@flannelhead 

One argument _against_ using _LAST enums, btw, might be seen with `DT_ADAPTATION_LAST`. It appears in _many_ switch statements (to satisfy careful compilers) but is never actually used for example to size an array (as far as I can tell).

flannelhead commented:
I will try to test this one in the weekend.

> This stuff is very tricky to test without "legacy images" and images that tick all the "special case" boxes

It should be possible to generate such legacy XMPs by temporarily removing the special case handling so that all the options are present in the combobox, then choosing them and saving a XMP. I have done this while testing the hiding of the legacy Euclidean norm in filmic.

> One argument against using _LAST enums, btw, might be seen with DT_ADAPTATION_LAST. It appears in many switch statements (to satisfy careful compilers) but is never actually used for example to size an array (as far as I can tell).

Indeed, I also had a look and had the same thoughts. That should probably be cleaned up, but it's not in the scope of this PR.

dterrahe commented:
> so that all the options are present in the combobox

True, but some of the options depend on the _type_ of image (xtrans etc). Not _impossible_ to test (at all) but _tedious_ :-) So needs to be done (too) by someone who's not this programmer (who is likely to make the same thinko both in coding _and_ testing).

> I will try to test this one in the weekend.

Thank you! No hurry, because:

> That should probably be cleaned up, but it's not in the scope of this PR.

I may split this PR in two; one for the trivial bug fix and one clean-up issues around this topic (that could be reverted if problems arise). Currently looking at denoiseprofile that can be streamlined now that the building blocks are more robust...

TurboGit commented on src/bauhaus/bauhaus.c:
```diff
@@ -1294,8 +1294,11 @@ gboolean dt_bauhaus_combobox_add_introspection(GtkWidget *widget,
   {
-    dt_bauhaus_combobox_add_full(widget, Q_(item->description ? item->description : item->name),
-                                 DT_BAUHAUS_COMBOBOX_ALIGN_RIGHT, GUINT_TO_POINTER(item->value), NULL, TRUE);
+    const char *text = item->description ? item->description : item->name;
```
ok, this is the main part, no description or name we don't add the combo entry.

TurboGit commented on src/bauhaus/bauhaus.c:
```diff
@@ -1552,6 +1555,16 @@ gboolean dt_bauhaus_combobox_set_from_value(GtkWidget *widget, int value)
     }
   }
+
+  dt_introspection_type_enum_tuple_t *values
```
But I don't get this fallback... If the item is not found in the loop above we come here...

Maybe some comments would help?

dterrahe replied:
This is really unrelated to the original bugreport (the solution of which is just the part above in dt_bauhaus_combobox_add_introspection).

Here, whenever an enum value is not currently in the combo (presumably because it is legacy and was manually removed in gui_init after the combo was created _from_param because we don't want _new_ images to use it), if its value is being selected (presumably from loading an old image that used the legacy value) we try to add it back to the combo from introspection . This means we don't have to do it manually in gui_update, which is messy. So filmic and highlights (and denoiseprofile) can be simplified.

TurboGit reviewed (commented):
For me this is a nice work. The code is simplified and less error prone. Yet I have one question about a part I don't understand.

dterrahe commented:
Now also simplified denoiseprofile.c. I found it hard to understand the (un)hiding of widgets based on selected options, since it happened in `gui_update`, `gui_changed` and also `mode_callback`. Since out-of-order enums and even hidden legacy options are now fully supported by `_from_params`, the separate `mode_callback` is no longer necessary and `gui_update` calls `gui_changed` like elsewhere. Hopefully I have transposed all settings correctly from one method to another and haven't missed subtle interactions. Thorough testing needed, but should be more maintainable going forward.

@TurboGit I've split out the pure bugfix into a standalone first commit. This could be cherrypicked in advance of more testing of the rest.

TurboGit reviewed (commented):
This looks good to me, I'll let some more time for others to review if needed as part are delicate. I haven't tested this last version. I'll do that soon.

dterrahe commented:
Since this turned into a cleanup of remaining introspection work I've also fully converted the lens correction module, since the specific workarounds it used are not needed anymore. The usual benefits; less duplication/shorter code, more descriptive history tooltips and ability to select specific dropdown items in shortcuts or from lua.

"drawback"; I've dropped the partial attempt to support lensfun versions < 0.2.6 (?). I'm somewhat doubtful it would have worked anyway, but let me know if this was an important feature that needs to be carried forward.

An issue this brought to my attention again is that I feel we don't do legacy support correctly. Every new module version in theory requires changes to conversions from _all_ previous versions (in addition to adding a new conversion from the current to the new version). This runs the risk of introducing errors in any of those old conversions that will likely not ever be tested if they are from a version older than the version in the regression test set. And it may make the results when skipping a version in the upgrade path different (A->C not necessarily equal to A->B->C). So I feel (but might be easily convince otherwise) that the correct way of doing this is to at each new module version introduction to _only_ add a legacy_params path from the current to the new version. This would require some infrastructure support to call legacy_params iteratively if the user has skipped a version. And obviously we would only want to do this going forward because "rewriting" all the existing conversions would be a very bad idea.

TurboGit commented:
> "drawback"; I've dropped the partial attempt to support lensfun versions < 0.2.6 (?).

At this point it is not an issue to me.

TurboGit commented:
@dterrahe : About the legacy params update that's how it is done for lib, I have never understood why it was not done for iop module, maybe I'm missing something. See routine `_update_params` in `lib.c`.

dterrahe commented:
> See routine `_update_params` in `lib.c`.

Yeah, that looks sensible. But for libs the `legacy_params` _returns_ the new version whereas for iops both old and new version are only inputs. So it would have to be a more complex polling routine to figure out which upgrade paths are supported.

`dt_masks_legacy_params` looks "interesting" as well.

After changing the supported process, we'll probably see people "copy/pasting" from the existing approach for a long time. Like I said, we wouldn't want to start messing with those existing codes because boring->errors and hard to test, but you'd have to police people submitting new upgrade paths using the old pattern and _require_ them to change it.

TurboGit commented:
Ok, I'm cherry-picking the first commit then.

flannelhead commented:
@dterrahe I didn't yet get to review this but will try again next weekend.

dterrahe commented:
Oh great! Didn't mean to put pressure on you; just thought if you didn't have time we could leave testing to the wider group...

flannelhead reviewed (changes requested):
Some notes in the inline comments.

I did a brief testing on the highlight reconstruction module as it was the most complicated case of these.

1. Took a couple of images: one with Bayer pattern and one xtrans.
2. Commented out the code that rebuilds the mode selection combobox and selected the legacy "reconstruct color" preset for both images.
3. Uncommented the menu rebuilding
4. Loaded both images and observed the options that were available in the mode selection:
With the Bayer image, all options were present. With the xtrans image, all options but guided laplacians were present.
5. Changed both images to "inpaint opposed"
6. Loaded them again in darkroom and observed that "reconstruct color" had vanished from the list.

The logic in the code seemed correct (as in doing the same thing as the original `highlights` code), and according to this test I would feel ok with merging this (after considering the little comments).

I concentrated my review effort on the couple of first commits `don't show nondescript (_LAST) enum in combo` and `simplify re-adding legacy options to combo` and only tested the `denoiseprofile` and `lens` modules very briefly.

Hope this helps.

dterrahe commented on src/iop/filmicrgb.c:
```diff
@@ -114,9 +114,8 @@ typedef enum dt_iop_filmicrgb_methods_type_t
-  DT_FILMIC_METHOD_EUCLIDEAN_NORM_V2 = 5, // $DESCRIPTION: "RGB euclidean norm"
   DT_FILMIC_METHOD_EUCLIDEAN_NORM_V1 = 4, // $DESCRIPTION: "RGB euclidean norm (legacy)"
-  DT_FILMIC_METHOD_LAST = 6, // this must always be greatest of the numbers above + 1
+  DT_FILMIC_METHOD_EUCLIDEAN_NORM_V2 = 5, // $DESCRIPTION: "RGB euclidean norm"
```
The practice is (seems to be...) to explicitly number all enums that are part of params for backward compatibility.

By putting DT_FILMIC_METHOD_EUCLIDEAN_NORM_V1 in the position corresponding to its number, it is possible to directly use that in `dt_bauhaus_combobox_remove_at`. Otherwise either we must assume it is always at the end (and we could use _LAST-1), or we need a new function `dt_bauhaus_combobox_remove_value`. I didn't think it was worth introducing that for the _very_ few cases that would use (and _test_) it.

I think this kind of stuff will always be somewhat messy and needs people to know exactly what they are doing, so I considered it not worth trying to make it "fool" proof (and encouraging people to use it thoughtlessly).

flannelhead replied:
That's good, thanks for the explanation.

dterrahe commented on src/bauhaus/bauhaus.c:
```diff
@@ -1294,8 +1294,11 @@ gboolean dt_bauhaus_combobox_add_introspection(GtkWidget *widget,
-    dt_bauhaus_combobox_add_full(widget, Q_(item->description ? item->description : item->name),
-                                 DT_BAUHAUS_COMBOBOX_ALIGN_RIGHT, GUINT_TO_POINTER(item->value), NULL, TRUE);
+    const char *text = item->description ? item->description : item->name;
+    if(*text)
```
`item->name` if the condition of the _for_ loop and `item->description` is the condition in the ternary operator and only used if not-null, so I believe they have both just been checked, but please double verify my reasoning and tell me if it's wrong.

Introspection will generate a `""` description, rather than a NULL if no $DESCRIPTION marker is found in the comment.

flannelhead replied:
Ah true, I didn't look broadly enough and missed the part where `item->name` is checked. Everything ok here.

dterrahe commented on src/iop/channelmixerrgb.c:
```diff
@@ -3855,12 +3855,8 @@ void reload_defaults(dt_iop_module_t *module)
-                                     GINT_TO_POINTER(DT_ILLUMINANT_CAMERA), NULL, TRUE);
-    }
+      // add back if needed
+      dt_bauhaus_combobox_set_from_value(g->illuminant, DT_ILLUMINANT_CAMERA);
```
`gui_update` and `dt_bauhaus_update_module` will be called after `reload_defaults` so that's where the value will be set to the user selected option.

I could explicitly mention that here, but I always struggle with how much commentary to put into the code base (since often we start from nothing). Putting the first paragraph in a comment in the code might lead to a situation where somebody changes the control flow and this is actually no longer true, but not updating comments and thereby creating more confusion. So in the end one needs to know quite a bit about the gui update processes to understand what is safe to touch. It took me a while...

BTW in this case an update to the combo value doesn't cascade to possible dependent processes because during `reload_defaults` `darktable.gui->reset` is TRUE.

flannelhead replied:
Fair enough.

Just expanding on this, would this warrant a `dt_bauhaus_combobox_add_from_value` which just adds the item from introspection (if missing) and does not change the actual set value? This would communicate the intent better here, removing the need for a separate explanatory comment.

dterrahe replied:
Then I might as well use `dt_bauhaus_combobox_add_introspection` which is more verbose but also more efficient.
I will need to add a check if the value currently is available (and return its position) and refactor a little using this `dt_bauhaus_combobox_get_from_value`; that will also make removing items (in gui_init) safer.

dterrahe replied:
@flannelhead Last commit not exhaustively tested yet, but would like your feedback on whether this actually improves readability or not...

flannelhead replied:
@dterrahe I left some comments now, let's see what you think of them. I'll let @TurboGit also comment on the readability side.

flannelhead commented:
It seems I can't resolve the discussions and @TurboGit will have to do it (kind of strange being able to start those discussions but not resolve them).

flannelhead commented on src/iop/channelmixerrgb.c:
```diff
@@ -3854,11 +3854,16 @@ void reload_defaults(dt_iop_module_t *module)
-      // add back if needed
-      dt_bauhaus_combobox_set_from_value(g->illuminant, DT_ILLUMINANT_CAMERA);
+    {
+      if(pos == -1)
```
you can combine this conditional to the parent `if`.

dterrahe replied:
I have to admit I almost did, but that would mean that the "else" would be executed whenever the DT_ILLUMINANT_CAMERA is there, removing it even if it should stay. Or am I misunderstanding you?

flannelhead replied:
Ah true! My mistake.

flannelhead commented on src/iop/channelmixerrgb.c:
```diff
@@ -3854,11 +3854,16 @@ void reload_defaults(dt_iop_module_t *module)
-      dt_bauhaus_combobox_set_from_value(g->illuminant, DT_ILLUMINANT_CAMERA);
+    {
+      if(pos == -1)
+        dt_bauhaus_combobox_add_introspection(g->illuminant, NULL,
```
Yes, this seems to communicate the intent a little bit better in my opinion.

flannelhead commented on src/iop/channelmixerrgb.c:
```diff
@@ -3854,11 +3854,16 @@ void reload_defaults(dt_iop_module_t *module)
+    {
+      if(pos == -1)
+        dt_bauhaus_combobox_add_introspection(g->illuminant, NULL,
+                                              module->so->get_f("illuminant")->Enum.values,
```
I liked the brevity of the previous
```
dt_bauhaus_combobox_set_from_value(g->illuminant, DT_ILLUMINANT_CAMERA);
```
- `dt_bauhaus_combobox_add_introspection` on the other hand seems a bit verbose because one has to dig up the enum tuple and pass it as well (plus the `NULL` action parameter). As a module author I wouldn't be able to very quickly pull the `module->so->get_f("illuminant")->Enum.values` unless looking at other modules. (It feels slightly like this is an "internal" structure that's not supposed to be used directly)

It would be convenient if there was be a variant of `dt_bauhaus_combobox_add_introspection` that looked up the enum tuple list itself, like `dt_bauhaus_combobox_set_from_value` does.

Obviously this is now just aesthetics and splitting hairs, and you may as well ignore this one.

dterrahe replied:
`dt_bauhaus_combobox_set_from_value` way of finding the enum is quite roundabout (first looking up the action in a hashtable and then looking up an associated enum list in another hashtable), whereas here we already "know" where to find the table. Agreed with the "internal structure" comment, but also don't necessarily want to normalise changing lists on the fly all the time either? And we already have ` module->get_f("illuminant")->Enum.Default;` and `Float.Default` in places, so if we start introducing helper functions for all of them...

If we _do_ want a helper function for this, it should just be to look up the `dt_introspection_type_enum_tuple_t` because I'd still want the flexibility of `dt_bauhaus_combobox_add_introspection` to be able to add a _series_ of values in one go and that function isn't _just_ used for introspection generated lists (even if they use that format).

flannelhead replied:
Fair enough, I don't think there's a need for further helper functions.

dterrahe replied:
If this pattern is used more in future (tbh people will learn how to do this by looking at other modules and copy/pasting anyway, no matter how obvious a function is) the need for a helper function/refactoring can always be reviewed. And at that point with more data on what common use patterns are.

Thanks again for reviewing and commenting. Much appreciated!

dterrahe commented:
@TurboGit just a kind reminder that I think this is ready to go.

TurboGit reviewed (approved):
Works for me, thanks!
